[flang] Add --gcc-toolchain and --gcc-install-dir options to flang.

Meinersbur opened the pull request:
The `--gcc-toolchain` and `--gcc-install-dir` option were previously only visible to the Clang driver, but not Flang. These determine which assembler, linker, and libraries to use, e.g. for cross-compiling, and therefore are relevant for Flang as well.

Fixes #86729

Meinersbur committed:
[flang] Add --gcc-toolchain and --gcc-install-dir options to flang

Meinersbur committed:
Combine --gcc-toolchain and -gcc-install-dir tests.

Also fix -help output checks.

tblah reviewed (commented):
Overall this looks good. Just minor comments.

Thank you for working on this!

Meinersbur committed:
Check gcc candidate installation search

banach-space commented on flang/test/Driver/gcc-toolchain-install-dir.f90:
```diff
@@ -0,0 +1,20 @@
+!! It does not test cross-compiling (--sysroot), so crtbegin.o, libgcc/compiler-rt, libc, libFortranRuntime, etc. are not supposed to be affected.
+
+! RUN: %flang 2>&1 -### -v -o %t %s -no-integrated-as -fuse-ld=ld --target=i386-unknown-linux-gnu --gcc-toolchain=%S/Inputs/basic_cross_linux_tree/usr                                         | FileCheck %s --check-prefix=CHECK-I386
+! RUN: %flang 2>&1 -### -v -o %t %s -no-integrated-as -fuse-ld=ld --target=i386-unknown-linux-gnu --gcc-install-dir=%S/Inputs/basic_cross_linux_tree/usr/lib/gcc/i386-unknown-linux-gnu/10.2.0 | FileCheck %s --check-prefix=CHECK-I386
```
[nit] The actual path doesn't really matter here, but IMHO something like `path/to/my/gcc-toolchain` instead of `Inputs/basic_cross_linux_tree/usr/lib/gcc/i386-unknown-linux-gnu/10.2.0` would be easier to parse for humans. Also, `Inputs/basic_cross_linux_tree/usr/lib/gcc/i386-unknown-linux-gnu/10.2.0` might indicate that somehow `gcc-10.2` is somehow special.

banach-space commented on flang/test/Driver/Inputs/basic_cross_linux_tree/usr/x86_64-unknown-linux-gnu/bin/ld.lld:
Why do we need these files?

Meinersbur replied:
I think the contents do not matter, but clangDriver checks for the existence of these files to validate a GCC installation. Clang's driver tests also have this content (some are symlinks to sibling files, but it doesn't seem to matter either).

banach-space replied:
> clangDriver checks for the existence of these files to validate a GCC installation

Even when using `-###`?

Meinersbur committed:
Fix Windows test

Meinersbur committed:
Add crtbegin.o to mock gcc installation

It is used by Flang/Clang driver to determine the validity of the installation.

Please do not add *.o to .gitignore

Meinersbur commented on flang/test/Driver/Inputs/basic_cross_linux_tree/usr/i386-unknown-linux-gnu/lib/.keep:
This is a copy of a mock GCC installation from Clang's driver test. It contains `.keep` file presumably because the clangDriver checks for the existence of this directory. Without at least one file in a directory, git will not create the directory when checking it out.

Meinersbur committed:
Fix missing slash in check line

Meinersbur committed:
Mark mock root binaries as executable

Meinersbur commented:
I could try to minimize the number of files/directories in basic_cross_linux_tree to be sufficient for only `--gcc-toolchain` and `--gcc-install-dir`, but I also think those would be useful for additional tests, e.g. such as `--use-ld`.

Meinersbur committed:
Add missing execute flag on mock executables

Meinersbur commented:
It was surprisingly difficult to make the test pass on buildkite while it worked locally. Mostly because `git add` by itself didn't replicate my worktree: The executable bits are not set and a `.gitignore` contained `*.o`. Very frustrating, but not it is showing green.

banach-space commented:
> It was surprisingly difficult to make the test pass on buildkite while it worked locally. Mostly because `git add` by itself didn't replicate my worktree: The executable bits are not set and a `.gitignore` contained `*.o`. Very frustrating, but not it is showing green.

Thanks for being so thorough with this, LGTM!

tblah reviewed (approved):
Thanks for sticking with this. LGTM

ye-luo reviewed (approved):
Tested. Works well for me.
